gnucash/report/reports/standard/investment-lots.scm

```diff
@@ -0,0 +1,2108 @@
+              ((get-long-term-sold-value) (lambda () long-term-sold-value))
+              ((get-short-term-realized-gain)
+                  (lambda () short-term-realized-gain))
+              ((get-long-term-realized-gain) (lambda () long-term-realized-gain))
+              ((get-end-amount) (lambda () end-amount))
+              ((get-end-basis) (lambda () end-basis))
+              ((get-end-value) (lambda () end-value))
+              ((get-unrealized-gain) (lambda () unrealized-gain))
+              ((get-short-term-unrealized-gain)
+                  (lambda () short-term-unrealized-gain))
+              ((get-long-term-unrealized-gain)
+                  (lambda () long-term-unrealized-gain))
+              ((get-has-warnings) (lambda () has-warnings))
+              ((get-is-active-in-window) (lambda () is-active-in-window))
+              ((get-currency) (lambda () currency))
+              (else (gnc:error (format #f "Bad action: ~a" (car args)))))
+            (cdr args)))))
+
+    ;; Create a stats collector for the grand total (all accounts) row.
+    (define all-accounts-info
+        (create-lot-stats-collector 0))
+
+    ;; Adds the specified account to the report.
+    (define (add-account-report account)
+      (let* ((currency (xaccAccountGetCommodity account))
+             (currency-fraction (gnc-commodity-get-fraction currency))
+             ;; Note that this price may not be in the report-currency. Use 
+             ;; (price->report-currency-value price) to convert.
+             (price (price-fn currency))           
+             (splits (get-all-splits account))
+             (get-lots-start-time-ms (current-milliseconds))
+             (lot-pair (get-all-lots splits))
+             (lots (car lot-pair))
+             (unassigned-splits (cadr lot-pair))
+             (get-lots-end-time-ms (current-milliseconds))
+
+             (get-account-balance (lambda()
+                (let ((unit-collector (gnc:account-get-comm-balance-at-date
+                                        account
+                                        to-date
+                                        #f))) ;; include-children?
+                  (cadr (unit-collector 'getpair currency #f)))))
+             ;; Gets whether the account should be included in the report,
+             ;; based on the account balance and the option to ignore empty
+             ;; accounts.
+             (check-account-balance (lambda() 
+                (or 
+                  include-empty-accounts
+                  (not (gnc-numeric-zero-p (get-account-balance))))))
+             ;; Gets whether the account should be included in the report,
+             ;; based on the number of lots and the option to ignore accounts
+             ;; with no lots.
+             (check-lot-count (lambda()
+                (or include-lotless-accounts
+                  (and 
+                    (not (null? lots))
+                    (> (length lots) 0))))))
```

Comment:
These two conditionals linked with `and` are identical. Plus, `length` is O(N). It's best to use a simple `(pair? lots)`.